src/file_data_source.cpp

```diff
@@ -113,6 +113,8 @@ const FileDataSource&FileDataSource::operator=(FileDataSource&&o){
 }
 
 void FileDataSource::open(const char*file_name){
+	if(file_descriptor != nullptr)
+		close();
```

Comment:
This leaves FileDataSource in an ill state if the new file cannot be opened and an exception is thrown. First open the new file and only if that succeeded, close the old one, and finally overwrite file_descriptor.